# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/layers/wr/WebRenderLayerManager.cpp
# Commit: f215196f1d1c
# Full Hash: f215196f1d1c8c6b42214b35fee351ef021bd704
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2020-03-28 09:32:56
# Description:
#   Bug 1624125 - Track display list changes in DisplayItemCache r=jrmuizel
#   
#   This is needed because display lists and DisplayItemCache have different lifetimes. For example, display lists can outlive WebRenderLayerManager when device reset occurs.
#   
#   A slightly nicer way of fixing this would be to couple DisplayItemCache with nsDisplayList or nsDisplayListBuilder. This is would currently require a lot of refactoring to look nice, because the painting code still supports non-retained display lists and non-WR code paths.
# ==============================================================================

diff -r 13a7265ed731 -r f215196f1d1c gfx/layers/wr/WebRenderLayerManager.cpp
--- a/gfx/layers/wr/WebRenderLayerManager.cpp	Fri Mar 27 16:49:40 2020 +0000
+++ b/gfx/layers/wr/WebRenderLayerManager.cpp	Fri Mar 27 16:49:37 2020 +0000
@@ -46,6 +46,14 @@
     mStateManagers[renderRoot].mRenderRoot = renderRoot;
     mStateManagers[renderRoot].mLayerManager = this;
   }
+
+  if (XRE_IsContentProcess() &&
+      StaticPrefs::gfx_webrender_enable_item_cache_AtStartup()) {
+    static const size_t kInitialCacheSize = 1024;
+    static const size_t kMaximumCacheSize = 10240;
+
+    mDisplayItemCache.SetCapacity(kInitialCacheSize, kMaximumCacheSize);
+  }
 }
 
 KnowsCompositor* WebRenderLayerManager::AsKnowsCompositor() { return mWrChild; }
@@ -198,6 +206,8 @@
     return false;
   }
 
+  mDisplayItemCache.SkipWaitingForPartialDisplayList();
+
   // Since we don't do repeat transactions right now, just set the time
   mAnimationReadyTime = TimeStamp::Now();
 
@@ -347,14 +357,12 @@
     // generating the WR display list is the closest equivalent
     PaintTelemetry::AutoRecord record(PaintTelemetry::Metric::Layerization);
 
-    builder.UpdateCacheState(aDisplayListBuilder->PartialBuildFailed());
+    mDisplayItemCache.SetDisplayList(aDisplayListBuilder, aDisplayList);
 
     mWebRenderCommandBuilder.BuildWebRenderCommands(
         builder, resourceUpdates, aDisplayList, aDisplayListBuilder,
         mScrollDatas, std::move(aFilters));
 
-    builder.UpdateCacheSize();
-
     builderDumpIndex =
         mWebRenderCommandBuilder.GetBuilderDumpIndex(builder.GetRenderRoot());
     containsSVGGroup = mWebRenderCommandBuilder.GetContainsSVGGroup();